add docs for mock improvements (#2958)
aws-amplify/docs: docs/cli/usage/mock.md

3. `AWS::Region`, `AWS::AccountID`, `AWS::StackName`, and `AWS::StackId` are resolved by parsing the `awscloudformation` configuration of the current environment in `team-provider-info.json`
4. Parameters constructed from depending on other resources in the project are resolved by parsing the `amplify-meta.json`. Additionally, if a mock API is currently running and the function depends on the API, the local API URL will be substituted for the cloud URL

In addition to using the context above to resolve CloudFormation template values, the mock environment will populate [lambda runtime environment variables](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime) in the following way:
1. `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN` will be populated using the AWS credentials that the CLI is currently configured to use
2. `_HANDLER`, `AWS_REGION`, `AWS_LAMBDA_FUNCTION_NAME`, `LAMBDA_TASK_ROOT`, and `LAMBDA_RUNTIME_DIR` will be set based on the function being mocked

> 3. Static defaults will be specified for all other runtime environment variables. The full list of static defaults can be found [here](https://github.com/aws-amplify/amplify-cli/blob/af76446d18bf626ca5f91c3ad41081175c959807/packages/amplify-util-mock/src/utils/lambda/populate-lambda-mock-env-vars.ts#L39)

Author: good catch!

If any of the default environment values do not work for your mocking use case, you can also specify overrides in a `.env` file within the function directory (ie. `<project root>/amplify/backend/function/<function name>/.env`).

### Connecting to a mock @model table

> If you have a function that connects to a table generated by @model you may want to test out this code against the mock table when running the mock API. To do that, create a `.env` file in the function directory and add the following:

Reviewer: ```suggestion
If you want to test a function to connects to a mocked table generated by @model, create a `.env` file in the function directory and add the following:
```
reword this, not happy with my wording either but current approach seems verbose.

Author: Reworded to "Connect a mock function that operates on a table generated by `@model` to the mock table when running `amplify mock api` by:"

Still seems a bit convoluted by it's at least a shorter sentence to grok

```
AWS_REGION=us-fake-1
DDB_ENDPOINT=http://localhost:62224
AWS_ACCESS_KEY_ID=fake
AWS_SECRET_ACCESS_KEY=fake
API_<api name>_<model>TABLE_NAME=<model>Table
```

> Ensure that you replace `<api name>` with the name of your API and `<model>` with the name of your model. The environment variable name should also be all caps. For example, if you have an API named "FlightStats" and a model defined as `type Airplane @model {...}` then then last line of the `.env` file should be:

Reviewer: ```suggestion
Replace `<api name>` with the name of your API and `<model>` with the name of your model. The environment variable name should be all capitalized. For example, if you have an API named "FlightStats" and a model defined as `type Airplane @model {...}`, then the last line of the `.env` file should be:
```

```
API_FLIGHTSTATS_AIRPLANETABLE_NAME=AirplaneTable
```

With the above `.env` configuration, you can then configure the DynamoDB call within the function as follows:
```javascript
const ddb = new aws.DynamoDB.DocumentClient({
    endpoint: process.env.DDB_ENDPOINT,
});
const result = await ddb.put({
    TableName: process.env.API_FLIGHTSTATS_AIRPLANETABLE_NAME,
    Item: {
        id: '1234567890',
        name: 'F-22',
        description: 'Cool fighter jet',
    },
}).promise();
```

> When running in the cloud, all of these environment variables will have valid values with the exception of `DDB_ENDPOINT` which will be undefined. In this case the DynamoDB client will use the default endpoint. When running locally, these environment variables will be populated using the local .env file which specifies the fake values required to connect to the local database. Also ensure that you run `amplify mock api` before running `amplify mock function` so that the local DynamoDB endpoint is active.

Reviewer: ```suggestion
When running in the cloud, all of these environment variables will have valid values except `DDB_ENDPOINT` which will be `undefined`. In this case the DynamoDB client will use the default endpoint. 

When running locally, these environment variables will be populated using the local .env file which specifies the fake values required to connect to the local database. 
```

"Also ensure that you run `amplify mock api` before running `amplify mock function` so that the local DynamoDB endpoint is active." should probably be worked in higher in the flow. This might be easier if we convert it to a "step-by-step" guide-like content.

Author: Converted this flow into numbered steps

> > Note: because the AWS credential environment variables have been overwritten with fake credentials in order to connect to the mock database, calls to other AWS resources when mocking will not work. If you want to call a mock table as well as other AWS services when mocking your function you will need to add some logic to switch between the fake credentials and the real ones in the DynamoDB client configuration. In this case, you could configure your `.env` file with:

Reviewer: ```suggestion
> Note: While connecting to the mock database, calls to other AWS resources when mocking will not work because the AWS credential environment variables have been overwritten with fake credentials. To call a mock table as well as other AWS services, add logic to switch between the fake credentials and the real ones in the DynamoDB client configuration. In this case, you could configure your `.env` file with:
```

Maybe shorten even more.

```
IS_MOCK=true
```
And configure the DynamoDB client with
```javascript
const clientConfig = process.env.IS_MOCK ? {
    region: 'us-fake-1',
    endpoint: 'http://localhost:62224',
    credentials: new aws.Credentials({
        accessKeyId: 'fake',
        secretAccessKey: 'fake',
    }),
} : undefined;
const ddb = new aws.DynamoDB.DocumentClient(clientConfig);
```

### Function mock limitations

> `amplify mock function` provides a great way to quickly get started testing your Lambda functions. However, it does not attempt to fully simulate the Lambda runtime environment and there may be some cases where the behavior of your function when mocking differs from executing in the cloud. For example, mock runs on your local OS and does not attempt to emulate Amazon Linux which is what your function will execute on in the cloud. Testing your function with mock should be used to get quick feedback on the correctness of your function but should not be used as a substitute for testing in a cloud development environment.

Reviewer: ```suggestion
`amplify mock function` does not attempt to fully simulate the Lambda runtime environment. There may be some cases where the behavior of your function when mocking differs from executing in the cloud. 

For example, mock runs on your local OS and does not attempt to emulate Amazon Linux which is what your function will execute on in the cloud. Testing your function with mock should be used to get quick feedback on the correctness of your function but should not be used as a substitute for testing in a cloud development environment.
```
